Add desktop notifications for MacOS

Dowsley opened the pull request:
# Desktop Notifications for macOS

## Summary
Partial implementation for https://github.com/godotengine/godot-proposals/issues/1338. Adds support for desktop notifications on macOS, allowing games and applications to send native system notifications even when the application is not in focus. Also supports scheduling them for future delivery.

## Changes
- Added `OS::send_notification()` method, that can be implemented in any system, with the following signature:
  ```gdscript
  Error send_notification(String title, String message, Callable callback = Callable(), Variant icon = null, int duration = -1, Dictionary datetime_to_send = {})
  ```
- Include MacOS implementation, which includes the following functionalities:
    - Custom Icons
    - Callbacks
    - Scheduling notifications
- Uses `UNUserNotificationCenter` on macOS 10.14+ and falls back to `NSUserNotification` on older versions
- Supports scheduling notifications for future delivery using the following datetime formats:
  - Full datetime: Dictionary with keys "year", "month", "day", "hour", "minute", "second"
  - Unix timestamp: Dictionary with key "unix_time"
  - Relative time: Dictionary with key "relative_seconds"
- Defensively handles edge cases like script mode, headless mode, and invalid app contexts
- Added documentation with examples
- Notifications work for Headless mode, but not for script mode (in which case, notification will be logged, and a warning will be shown).

## Current Limitations
- Notifications will only work when the application is properly exported/bundled (this is a macOS system limitation).
  - If the app is not bundled, notification will at least be logged along with a warning.
- Notifications will only show if application is out of focus.
- The duration parameter is not supported on macOS, as notification duration is controlled by the system's accessibility settings.

## Testing
1. Compile with `scons dev_build=yes generate_bundle=yes`. It will need to be bundled for notifications to work.
2. Download [these scripts](https://gist.github.com/Dowsley/ce2df692c1aa0f6a4a671f49c721bcf0)
3. Run one of the test scripts from the project: `./bin/godot_macos_editor_dev.app/Contents/MacOS/Godot --path /PATH/TO/PROJECT -s test_basic_notification.gd`
    - Each of the 4 scripts is made to test a specific functionality.
    - When you run the script, it will open a window. Since notifications only appear when app is out of focus, they have 2-3 seconds timers for you to click out of the new window to make it lose focus. Otherwise no notification will be shown.
5. If notification does not work even after following steps correctly, you need to go to System Settings > Notifications > GodotTest (or whatever the name of your bundle) > Enable Notifications

### Evidence/Live testing:
#### 1. Basic Notification
https://github.com/user-attachments/assets/74ff7d81-cb6e-4e27-8014-147ff79f7a6b

#### 2. Icons
https://github.com/user-attachments/assets/21fadcf1-344d-400e-9daf-528c7b086ea5

#### 3. Callbacks
https://github.com/user-attachments/assets/e2564bea-193a-433f-8194-4c8c2390508b

#### 4. Scheduling (minor correction: it was 10 seconds from the time it was sent, not 15)
https://github.com/user-attachments/assets/94176e20-9358-4b81-812e-a2748197c04e

#### 5. Any of the tests above, but executed in an unbundled project
![image](https://github.com/user-attachments/assets/33d5f0f5-76f9-4347-b94b-6e593744e270)

Also, while you're recording, notifications wont show in MacOS by default. Make sure to change this setting when recording to show me any possible bug.

## What can improve right now
- Did not have an opportunity to test feature on older MacOS versions. Would love some help.

## Future Work
- Investigate a possible implementation for duration.
- Implementation for other platforms, of course.

Dowsley commented on platform/macos/os_macos.mm:
```diff
@@ -944,6 +945,70 @@
 	[NSApp registerUserInterfaceItemSearchHandler:delegate];
 }
 
+Error OS_MacOS::send_notification(const String &p_title, const String &p_message, const Callable &p_callback, const Ref<Image> &p_icon, int p_duration) {
```
INFO: This is the most critical part and needs the most feedback. Left some comments to make my reasoning clear.

bruvzg commented:
> (MANUAL BUNDLE) Create simple app bundle wrapper:

`generate_bundle=yes` flag can be added to the scons build command to automatically bundle Godot. Templates for the bundles (for both editor and export) are in: https://github.com/godotengine/godot/tree/master/misc/dist

Dowsley commented:
> > (MANUAL BUNDLE) Create simple app bundle wrapper:
> 
> `generate_bundle=yes` flag can be added to the scons build command to automatically bundle Godot. Templates for the bundles (for both editor and export) are in: https://github.com/godotengine/godot/tree/master/misc/dist

Thanks! Updated the description.

Also, I'll work on fixing some failing pipelines - still getting the hang of the process, as well as implement the remaining functionalities. Meanwhile would be amazing to get feedback of the overral design.

Dowsley commented:
Added icon and duration support + updated videos. Updated description.

Calinou commented:
> This is fully expected, bundled app is the only supported way to run it on macOS.

Is it possible to symlink a binary to the macOS template bundle folder's as `Contents/MacOS/Godot`? This way, you can keep recompiling without `generate_bundle=yes` and it will run as bundled if you run the binary at the symlinked location.

Dowsley commented:
Still developing this. Managed to add support for callback. Before I commit need to address one concern about `p_duration`:

I believe MacOS only supports scheduling notifications, but not setting their duration. At least I couldn't find anything like this via `UNUserNotificationCenter`. If confirmed this is not supported, we will need to document that `p_duration` will not be taken into account for MacOS platform. I also proposed we throw a warning if caller passes non-default argument:

```
 if (p_duration > DEFAULT_VAL) {
     WARN_PRINT("Custom notification duration is not supported on macOS. System defaults will be used.");
 }
```

And regarding scheduling, I'm thinking, should we also consider adding a new parameter to the signature of `OS::send_notification` to support scheduling? Wondering what would be the optimal type of input for this argument to be easily used in any platform.

cc @bruvzg @Calinou

bruvzg commented:
> I believe MacOS only supports scheduling notifications, but not setting their duration

I do not think there's `duration` is directly supported, but it should be possible to periodically check `UNNotification.date` of delivered notification (`getDeliveredNotificationsWithCompletionHandler`) and delete them with `removeDeliveredNotificationsWithIdentifiers:`.

Dowsley commented:
> > I believe MacOS only supports scheduling notifications, but not setting their duration
> 
> I do not think there's `duration` is directly supported, but it should be possible to periodically check `UNNotification.date` of delivered notification (`getDeliveredNotificationsWithCompletionHandler`) and delete them with `removeDeliveredNotificationsWithIdentifiers:`.

I see. This seems somewhat hacky, and in such case, it might be worth leaving `duration` as unsupported for MacOS in this initial iteration. After all, the intended use of Banners is that their duration are automatically managed by the system and accessibility settings.

I think scheduling is a must however (even if it wasn't on the original proposal). So I added it.

Also added aforementioned changes.

Updated description to reflect current state.

Dowsley committed:
Add desktop notifications for MacOS

Dowsley commented:
Ready for first round of reviews

Mickeon commented on doc/classes/OS.xml:
```diff
@@ -773,6 +773,65 @@
+			<param index="5" name="datetime_to_send" type="Dictionary" default="{}" />
+			<description>
+				Sends a desktop notification to the user with the given title and message.
+
```
Every new line creates a "paragraph" as far as the class reference is concerned. Empty new lines are not necessary and will likely cause the static checks to fail.

Mickeon commented on doc/classes/OS.xml:
```diff
@@ -773,6 +773,65 @@
+				  - Unix timestamp: A single key "unix_time" with a Unix timestamp value.
+				  - Relative time: A single key "relative_seconds" specifying seconds from now when the notification should be sent.
+
+				[b]Note:[/b] This method is currently only implemented on macOS.
```
Let's not make promises.
```suggestion
				[b]Note:[/b] This method is only implemented on macOS.
```

Mickeon commented on doc/classes/OS.xml:
```diff
@@ -773,6 +773,65 @@
+				The notification behavior and appearance varies depending on the platform, with some platforms potentially ignoring certain parameters:
+				- [param title]: The title of the notification.
+				- [param message]: The main body text of the notification.
+				- [param callback]: Callable to be called when the user interacts with the notification (e.g., clicks on it). Note that not all platforms support this feature.
```
Unsure why the "note" is necessary here.
```suggestion
				- [param callback]: [Callable] to be called when the user interacts with the notification, such as by clicking it.
```

Mickeon commented on doc/classes/OS.xml:
```diff
@@ -773,6 +773,65 @@
+
+				The notification behavior and appearance varies depending on the platform, with some platforms potentially ignoring certain parameters:
+				- [param title]: The title of the notification.
+				- [param message]: The main body text of the notification.
```
```suggestion
				- [param message]: The main text of the notification.
```
or
```suggestion
				- [param message]: The body text of the notification.
```

Mickeon commented on doc/classes/OS.xml:
```diff
@@ -773,6 +773,65 @@
+			<description>
+				Sends a desktop notification to the user with the given title and message.
+
+				The notification behavior and appearance varies depending on the platform, with some platforms potentially ignoring certain parameters:
```
Very weird to begin by mentioning the behaviour of notifications per platform, then mentioning parameters being ignored, and then proceeding with a list. It implies that the list should be about which parameters are ignored and when, but it does not. 
I have no suggestions.

Mickeon commented on doc/classes/OS.xml:
```diff
@@ -773,6 +773,65 @@
+				- [param title]: The title of the notification.
+				- [param message]: The main body text of the notification.
+				- [param callback]: Callable to be called when the user interacts with the notification (e.g., clicks on it). Note that not all platforms support this feature.
+				- [param icon]: A custom icon to display with the notification. Pass an Image object or null to use the default application icon.
```
```suggestion
				- [param icon]: A custom icon to display with the notification. If [code]null[/code], the default application icon is used.
```

Mickeon commented on doc/classes/OS.xml:
```diff
@@ -773,6 +773,65 @@
+				- [param message]: The main body text of the notification.
+				- [param callback]: Callable to be called when the user interacts with the notification (e.g., clicks on it). Note that not all platforms support this feature.
+				- [param icon]: A custom icon to display with the notification. Pass an Image object or null to use the default application icon.
+				- [param duration]: How long the notification should remain visible in seconds. If set to -1 (default), the system default duration will be used. If set to 0, the notification will remain until manually dismissed (not supported on all platforms).
```
```suggestion
				- [param duration]: How long the notification should remain visible, in seconds. If set to [code]-1[/code] (default), the system default duration will be used. If set to [code]0[/code], the notification will remain until manually dismissed (not supported on all platforms).
```

Mickeon commented on doc/classes/OS.xml:
```diff
@@ -773,6 +773,65 @@
+				- [param callback]: Callable to be called when the user interacts with the notification (e.g., clicks on it). Note that not all platforms support this feature.
+				- [param icon]: A custom icon to display with the notification. Pass an Image object or null to use the default application icon.
+				- [param duration]: How long the notification should remain visible in seconds. If set to -1 (default), the system default duration will be used. If set to 0, the notification will remain until manually dismissed (not supported on all platforms).
+				- [param datetime_to_send]: A Dictionary that specifies when to schedule the notification. If empty (default), the notification is sent immediately. The Dictionary can be in one of these formats:
+				  - Full datetime: Must include keys "year", "month", "day", "hour", "minute", and "second".
+				  - Unix timestamp: A single key "unix_time" with a Unix timestamp value.
+				  - Relative time: A single key "relative_seconds" specifying seconds from now when the notification should be sent.
```
A list within a list is unique in the class reference. Can't say I have seen it anywhere, and I can't guarantee it'll look good in the built-in and online class reference. I would suggest compressing to a single line.

Mickeon commented on doc/classes/OS.xml:
```diff
@@ -773,6 +773,65 @@
+				  - Unix timestamp: A single key "unix_time" with a Unix timestamp value.
+				  - Relative time: A single key "relative_seconds" specifying seconds from now when the notification should be sent.
+
+				[b]Note:[/b] This method is currently only implemented on macOS.
+
+				[b]Note:[/b] On macOS, this implementation uses the UserNotifications framework (UNUserNotificationCenter) on macOS 10.14+ and falls back to NSUserNotification on older versions. On macOS 10.14+, notifications are automatically grouped together when multiple notifications are sent. When running in headless mode or with --script, the notification will be printed to the console instead.
+
+				[b]Note:[/b] The [param duration] parameter is not supported on macOS, the notification duration is controlled by the system's accessibility settings.
```
Ideally these notes should be moved all the way to the bottom below the examples.
Maybe not for the "The [param duration] parameter" note? Since it's more topical to the list of parameters above.

Mickeon commented on doc/classes/OS.xml:
```diff
@@ -773,6 +773,65 @@
+				[b]Example:[/b] Sending an immediate notification:
+				[codeblocks]
+				[gdscript]
+				func _on_operation_completed() -&gt; void:
```
No static typing unless necessary.
```suggestion
				func _on_operation_completed():
```
(Not unless we agree on how to do it)

Mickeon commented on doc/classes/OS.xml:
```diff
@@ -773,6 +773,65 @@
+				[b]Example:[/b] Scheduling a notification for 10 minutes from now:
+				[codeblocks]
+				[gdscript]
+				func schedule_reminder() -&gt; void:
+				    var datetime = {"relative_seconds": 600}  # 10 minutes
```
```suggestion
				func schedule_reminder():
				    var datetime = { "relative_seconds": 600 } # 10 minutes.
```

Mickeon commented on doc/classes/OS.xml:
```diff
@@ -773,6 +773,65 @@
+				[csharp]
+				private void ScheduleReminder()
+				{
+				    var datetime = new Godot.Collections.Dictionary { {"relative_seconds", 600} };  // 10 minutes
```
```suggestion
				    var datetime = new Godot.Collections.Dictionary { { "relative_seconds", 600 } }; // 10 minutes.
```

Mickeon reviewed (commented):
A lot of future-proofing in the description. I am not sure why.

I cannot speak on how useful this feature is.
